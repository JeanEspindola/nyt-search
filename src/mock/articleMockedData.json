{
  "responseList": {
    "response":
      {"docs":
        [
          {
            "web_url":"https://www.nytimes.com/2018/02/27/sports/usa-gymnastics-team.html",
            "snippet":"In the wake of a sex abuse scandal, many elite gymnasts fear the reckoning, essential though it is, may put their lifelong goals in jeopardy.",
            "headline": {
              "main":"U.S. Gymnasts Worry That Program’s Chaos Will Ruin Their Chances",
              "print_headline":"American Gymnasts Worry That a Program in Chaos Will Spoil Their Chances"
            },
            "pub_date":"2018-02-27T11:15:25+0000"
          },
          {
            "web_url":"https://www.nytimes.com/2018/03/01/sports/larry-nassar-gymnastics.html",
            "snippet":"Gymnasts, coaches, parents and fans respond to our question: Where does gymnastics go after the trial of Dr. Larry Nassar?",
            "headline": {
              "main": "The ‘Gymternet’ Reacts: Where Does Gymnastics Go From Here?",
              "print_headline": "Where Does the Sport Go From Here? The ‘Gymternet’ Reacts"
            },
            "pub_date":"2018-03-01T20:07:23+0000"
          },
          {
            "web_url":"https://www.nytimes.com/2018/02/26/arts/television/unsolved-murders-tupac-notorious-big-review.html",
            "snippet":"A 10-part limited series on USA is a lightly fictionalized account of the various investigations into the two rappers’ deaths two decades ago.",
            "headline": {
              "main":"Review: Who Killed Biggie and Tupac? ‘Unsolved’ Might Know",
              "print_headline":"Hip-Hop Murder Ballad"
            },
            "pub_date":"2018-02-26T18:10:37+0000"
          },
          {
            "web_url":"https://www.nytimes.com/2018/02/08/sports/olympic-swimmer-ariana-kukors-coach-abuse.html",
            "snippet":"Ms. Kukors, who competed in the 2012 Games, said that Sean Hutchison began sexually abusing her when she was 16. He denied the accusations.",
            "headline": {
              "main":"Olympic Swimmer Ariana Kukors Accuses Former Coach of Sexual Abuse",
              "print_headline":"Swimmer Accuses Former Coach of Abuse"
            },
            "pub_date":"2018-02-09T02:21:41+0000"
          }
        ]
      }
  },
  "formattedList": [
    {
      "id": 1,
      "web_url":"https://www.nytimes.com/2018/02/27/sports/usa-gymnastics-team.html",
      "snippet":"In the wake of a sex abuse scandal, many elite gymnasts fear the reckoning, essential though it is, may put their lifelong goals in jeopardy.",
      "headline": {
        "main":"U.S. Gymnasts Worry That Program’s Chaos Will Ruin Their Chances",
        "print_headline":"American Gymnasts Worry That a Program in Chaos Will Spoil Their Chances"
      },
      "pub_date":"2018-02-27T11:15:25+0000"
    },
    {
      "id": 2,
      "web_url":"https://www.nytimes.com/2018/03/01/sports/larry-nassar-gymnastics.html",
      "snippet":"Gymnasts, coaches, parents and fans respond to our question: Where does gymnastics go after the trial of Dr. Larry Nassar?",
      "headline": {
        "main": "The ‘Gymternet’ Reacts: Where Does Gymnastics Go From Here?",
        "print_headline": "Where Does the Sport Go From Here? The ‘Gymternet’ Reacts"
      },
      "pub_date":"2018-03-01T20:07:23+0000"
    },
    {
      "id": 3,
      "web_url":"https://www.nytimes.com/2018/02/26/arts/television/unsolved-murders-tupac-notorious-big-review.html",
      "snippet":"A 10-part limited series on USA is a lightly fictionalized account of the various investigations into the two rappers’ deaths two decades ago.",
      "headline": {
        "main":"Review: Who Killed Biggie and Tupac? ‘Unsolved’ Might Know",
        "print_headline":"Hip-Hop Murder Ballad"
      },
      "pub_date":"2018-02-26T18:10:37+0000"
    },
    {
      "id": 4,
      "web_url":"https://www.nytimes.com/2018/02/08/sports/olympic-swimmer-ariana-kukors-coach-abuse.html",
      "snippet":"Ms. Kukors, who competed in the 2012 Games, said that Sean Hutchison began sexually abusing her when she was 16. He denied the accusations.",
      "headline": {
        "main":"Olympic Swimmer Ariana Kukors Accuses Former Coach of Sexual Abuse",
        "print_headline":"Swimmer Accuses Former Coach of Abuse"
      },
      "pub_date":"2018-02-09T02:21:41+0000"
    }
  ]
}