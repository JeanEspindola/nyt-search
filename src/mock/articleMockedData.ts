export const responseList = {
  response: {
    docs: [
      {
        webUrl: 'https://www.nytimes.com/2018/02/27/sports/usa-gymnastics-team.html',
        snippet:
          'In the wake of a sex abuse scandal, many elite gymnasts fear the reckoning, essential though it is, may put their lifelong goals in jeopardy.',
        headline: {
          main: 'U.S. Gymnasts Worry That Program’s Chaos Will Ruin Their Chances',
          printHeadline: 'American Gymnasts Worry That a Program in Chaos Will Spoil Their Chances',
        },
        pubDate: '2018-02-27T11:15:25+0000',
      },
      {
        webUrl: 'https://www.nytimes.com/2018/03/01/sports/larry-nassar-gymnastics.html',
        snippet:
          'Gymnasts, coaches, parents and fans respond to our question: Where does gymnastics go after the trial of Dr. Larry Nassar?',
        headline: {
          main: 'The ‘Gymternet’ Reacts: Where Does Gymnastics Go From Here?',
          printHeadline: 'Where Does the Sport Go From Here? The ‘Gymternet’ Reacts',
        },
        pubDate: '2018-03-01T20:07:23+0000',
      },
      {
        webUrl: 'https://www.nytimes.com/2018/02/26/arts/television/unsolved-murders-tupac-notorious-big-review.html',
        snippet:
          'A 10-part limited series on USA is a lightly fictionalized account of the various investigations into the two rappers’ deaths two decades ago.',
        headline: {
          main: 'Review: Who Killed Biggie and Tupac? ‘Unsolved’ Might Know',
          printHeadline: 'Hip-Hop Murder Ballad',
        },
        pubDate: '2018-02-26T18:10:37+0000',
      },
      {
        webUrl: 'https://www.nytimes.com/2018/02/08/sports/olympic-swimmer-ariana-kukors-coach-abuse.html',
        snippet:
          'Ms. Kukors, who competed in the 2012 Games, said that Sean Hutchison began sexually abusing her when she was 16. He denied the accusations.',
        headline: {
          main: 'Olympic Swimmer Ariana Kukors Accuses Former Coach of Sexual Abuse',
          printHeadline: 'Swimmer Accuses Former Coach of Abuse',
        },
        pubDate: '2018-02-09T02:21:41+0000',
      },
    ],
  },
}

export const formattedList = [
  {
    id: 1,
    webUrl: 'https://www.nytimes.com/2018/02/27/sports/usa-gymnastics-team.html',
    snippet:
      'In the wake of a sex abuse scandal, many elite gymnasts fear the reckoning, essential though it is, may put their lifelong goals in jeopardy.',
    headline: {
      main: 'U.S. Gymnasts Worry That Program’s Chaos Will Ruin Their Chances',
      printHeadline: 'American Gymnasts Worry That a Program in Chaos Will Spoil Their Chances',
    },
    pubDate: '2018-02-27T11:15:25+0000',
  },
  {
    id: 2,
    webUrl: 'https://www.nytimes.com/2018/03/01/sports/larry-nassar-gymnastics.html',
    snippet:
      'Gymnasts, coaches, parents and fans respond to our question: Where does gymnastics go after the trial of Dr. Larry Nassar?',
    headline: {
      main: 'The ‘Gymternet’ Reacts: Where Does Gymnastics Go From Here?',
      printHeadline: 'Where Does the Sport Go From Here? The ‘Gymternet’ Reacts',
    },
    pubDate: '2018-03-01T20:07:23+0000',
  },
  {
    id: 3,
    webUrl: 'https://www.nytimes.com/2018/02/26/arts/television/unsolved-murders-tupac-notorious-big-review.html',
    snippet:
      'A 10-part limited series on USA is a lightly fictionalized account of the various investigations into the two rappers’ deaths two decades ago.',
    headline: {
      main: 'Review: Who Killed Biggie and Tupac? ‘Unsolved’ Might Know',
      printHeadline: 'Hip-Hop Murder Ballad',
    },
    pubDate: '2018-02-26T18:10:37+0000',
  },
  {
    id: 4,
    webUrl: 'https://www.nytimes.com/2018/02/08/sports/olympic-swimmer-ariana-kukors-coach-abuse.html',
    snippet:
      'Ms. Kukors, who competed in the 2012 Games, said that Sean Hutchison began sexually abusing her when she was 16. He denied the accusations.',
    headline: {
      main: 'Olympic Swimmer Ariana Kukors Accuses Former Coach of Sexual Abuse',
      printHeadline: 'Swimmer Accuses Former Coach of Abuse',
    },
    pubDate: '2018-02-09T02:21:41+0000',
  },
]
