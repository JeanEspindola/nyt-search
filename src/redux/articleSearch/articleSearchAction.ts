import {
  ArticleSearchActionTypes,
  ArticleSearchSubmitAction,
  ArticleSearchSubmitFailedAction,
  ArticleSearchSubmitInProgressAction,
  ArticleSearchSubmitSuccessAction,
  ArticleSearchSubmitType,
  ArticleSearchSuccessType,
} from './articleSearchTypes'

export const articleSearchSubmit = (
  payload: ArticleSearchSubmitType
): ArticleSearchSubmitAction => ({
  type: ArticleSearchActionTypes.SEARCH_SUBMIT_INIT,
  payload,
})

export const articleSearchSubmitInProgress = (): ArticleSearchSubmitInProgressAction => ({
  type: ArticleSearchActionTypes.SEARCH_SUBMIT_INPROGRESS,
})

export const articleSearchSubmitFailed = (): ArticleSearchSubmitFailedAction => ({
  type: ArticleSearchActionTypes.SEARCH_SUBMIT_FAILED,
})

export const articleSearchSubmitSuccess = (payload: ArticleSearchSuccessType) => ({
  type: ArticleSearchActionTypes.SEARCH_SUBMIT_SUCCESS,
  payload,
})

export type ArticleSearchActions =
  | ArticleSearchSubmitAction
  | ArticleSearchSubmitInProgressAction
  | ArticleSearchSubmitSuccessAction
  | ArticleSearchSubmitFailedAction
