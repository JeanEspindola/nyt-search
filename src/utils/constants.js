export const API_KEY = 'de99f27226cb42b1993b11f717641745';
export const URL = 'https://api.nytimes.com/svc/search/v2/articlesearch.json';

export const SEARCH_SUBMIT_SUCCESS = 'SEARCH_SUBMIT_SUCCESS';
