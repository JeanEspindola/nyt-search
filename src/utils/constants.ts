export const API_KEY = 'aaassdfdfdfdfdfdf12343KApkAJmTE896X2'
export const URL = 'https://api.nytimes.com/svc/search/v2/articlesearch.json'
export const BASE_URL = `${URL}?api-key=${API_KEY}`

/* Filtering the search for only return articles */
export const FILTER_QUERY = 'document_type:("article")'
/* Restricts the response fields to only values necessary */
export const RESPONSE_FIELDS = 'snippet,pub_date,headline,web_url'
