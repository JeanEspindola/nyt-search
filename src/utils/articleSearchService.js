import 'whatwg-fetch';
import { URL, API_KEY, FILTER_QUERY, RESPONSE_FIELDS } from './constants';

const baseUrl = `${URL}?api-key=${API_KEY}`;

class ArticleSearchService {
  static getSearchResults(query, page) {
    const queryParams = `&q=${query}&page=${page}&fq=${FILTER_QUERY}&fl=${RESPONSE_FIELDS}`;
    const url = `${baseUrl}${queryParams}`;

    return fetch(url)
      .then(response => response.json())
      .then(json => json.response.docs)
      .catch(error => error);
  }
}

export default ArticleSearchService;
