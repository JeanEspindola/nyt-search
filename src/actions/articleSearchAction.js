import articleSearchService from '../utils/articleSearchService';
import { SEARCH_SUBMIT_SUCCESS } from '../utils/constants';

export function searchQuerySuccess(list) {
  return {
    type: SEARCH_SUBMIT_SUCCESS,
    list,
  };
}

export function loadSearchResults(query, page) {
  return dispatch => articleSearchService.getSearchResults(query, page).then((articleList) => {
    dispatch(searchQuerySuccess(articleList));
  }).catch((error) => {
    throw (error);
  });
}
