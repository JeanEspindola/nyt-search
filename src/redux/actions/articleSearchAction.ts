/*
 * Article Search Action:
 *
 * All actions taken from components in order to access/modify the store.
 * In this case there is single action, when searching query success.
 *
 * */
import articleSearchService from '../../utils/articleSearchService'
import { SEARCH_SUBMIT_SUCCESS, LOADING_INDICATOR_SUCCESS } from '../../utils/constants'

export const loadingSubmit = (loading: boolean) => ({
  type: LOADING_INDICATOR_SUCCESS,
  loading,
})

export const searchQuerySuccess = (list: Array<never>, page: number, query: string) => ({
  type: SEARCH_SUBMIT_SUCCESS,
  list,
  page,
  query,
})

// @ts-ignore
export const loadSearchResults = (query: string, page: number) => async (dispatch) => {
  dispatch(loadingSubmit(true))

  try {
    const articleList = await articleSearchService.getSearchResults(query, page)
    dispatch(loadingSubmit(false))
    dispatch(searchQuerySuccess(articleList, page, query))
  } catch (error) {
    dispatch(loadingSubmit(false))
    throw error
  }
}
