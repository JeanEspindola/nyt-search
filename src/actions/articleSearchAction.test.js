import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { SEARCH_SUBMIT_SUCCESS, LOADING_INDICATOR_SUCCESS } from '../utils/constants';
import { searchQuerySuccess, loadingSubmit, loadSearchResults } from './articleSearchAction';
import articleMockedData from '../mock/articleMockedData.json';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('articleSearchAction', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });

  it('creates a LOADING_INDICATOR_SUCCESS action', () => {
    expect(searchQuerySuccess([], 1, 'test')).toEqual({
      type: SEARCH_SUBMIT_SUCCESS,
      list: [],
      page: 1,
      query: 'test',
    });
    expect(searchQuerySuccess([], 1, true)).toMatchSnapshot();
  });

  it('creates a SEARCH_SUBMIT_SUCCESS action', () => {
    expect(loadingSubmit(false)).toEqual({
      type: LOADING_INDICATOR_SUCCESS,
      loading: false,
    });
    expect(loadingSubmit(false)).toMatchSnapshot();
  });

  it('should handle loadSearchResults action - success', async () => {
    const response = articleMockedData.responseList;

    fetch.mockResponseSuccess(response);
    await store.dispatch(loadSearchResults('test', 1));
    expect(store.getActions()).toMatchSnapshot();
  });
});
