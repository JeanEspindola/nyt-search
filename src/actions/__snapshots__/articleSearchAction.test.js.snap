// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleSearchAction creates a LOADING_INDICATOR_SUCCESS action 1`] = `
Object {
  "list": Array [],
  "page": 1,
  "query": true,
  "type": "SEARCH_SUBMIT_SUCCESS",
}
`;

exports[`articleSearchAction creates a SEARCH_SUBMIT_SUCCESS action 1`] = `
Object {
  "loading": false,
  "type": "LOADING_INDICATOR_SUCCESS",
}
`;

exports[`articleSearchAction should handle loadSearchResults action - error 1`] = `
Array [
  Object {
    "loading": true,
    "type": "LOADING_INDICATOR_SUCCESS",
  },
]
`;

exports[`articleSearchAction should handle loadSearchResults action - success 1`] = `
Array [
  Object {
    "loading": true,
    "type": "LOADING_INDICATOR_SUCCESS",
  },
]
`;
